### DISCUSSION TITLE
Upcoming Changes

### DISCUSSION THREAD

**Discussion by reinhard-brandstaedter** [reinhard-brandstaedter/solarflow-control#144] [Announcements]
### First things first
In the past few weeks we have seen quite some feedback, questions and also one or the other issue reported. I want to thank everyone for their contribution, questions and observations!
With limited testing and setup possibilities (who owns more than one SF, with all the different depending devices, smartmeter reader, DTUs, ...) it is sometimes quite difficult to make everything work from the start, so issues are expected ;-)
Thanks for your patience!

### Upcoming Changes
Apart from all the feedback I've been also working on a change which hopefully makes the whole solarflow-control a bit more elegant and precise when it comes to steering. I still want to give it some testing but here is a summary of the changes about to come:

1. _Switching to a event-driven steering instead of a continuous loop_
Instead of updating limit settings and taking decision in a loop, solarflow-control will now only trigger those actions if there is a reason for that. Meaning either there is significant change in Solar Input and/or significant change in household demand. This approach should guarantee faster adoption to demand/production changes and at the same time avoid unnecessary operation.

1. _Linear regression forecasting of demand/production_
Instead of using weighted averages on source parameters like solarinput and household demand, we will now use a linear regression forecast of these values (based on the past). These forecasted values are then used to trigger the beforementioned event-driven steering

1. _Rapid change detection_
When big consumers are turned on/off we need to adjust any limits rather quick. This is not always easy as a quick change can also lead to noise/too many limit actions (especially when SF reacts rather slow), on the other hand when reacting too slow on the change it can mean that the power is either provided to late or is provided for too long (wasted).
Therefore we introduced a rapid change detection that clears the history of measurements and should (also with linear regression) lead to faster reaction times (however can't accelerate the hub's switching time)

1. _Detection of Inverter max capacity_
We have built in an detection algorithm for the maximum inverter capacity, based on current absolute and relative limits read from the DTU. This capacity should help in avoiding over-steering without configuration needs. In the future this might also help to adjust DTU/Inverter limits faster and more precisely.

1. _Location detection_
Using a different Geo IP API to determine the position and thus sunrise/sunset times. If solarflow-control doesn't have access to the internet no location detection will be possible (as before).

### Testing
If you already want to test these new features feel free to pull the _solarflow-control:dev_ docker image. Feedback is welcome!

**Comment by mavo** [comment]
Hi @reinhard-brandstaedter,
thank you for the explanation. Some days ago, before finding this announcement, I actually wanted to ask if there is something like a "roadmap" for planned features. Good I did not and instead found this.

I have a question to mentioned subject 5. Is it still possible to set geo location via coordinates as it is right now?

Also regarding the demand and production forecast, how will this actually work? Will the app "train itself" to the households usage patterns?

Again thx a lot for all your work, keep it up!
MaVo

**Reply by reinhard-brandstaedter** [reply]
Hi @mavo ,

regarding the geolocation detection this hasn't changed and manual location settings are still possible.

The demand / production forecast isn't as smart and doesn't use any seasonal learning. I first thought of that, but at the short timeframes we are operating there is no sense to do that. What the forecast actually does is using a linear regression based on a few measurements back and using these instead of moving averages.
Btw. so far - in combination with event based steering trigger theis seems to work quite well.

It's all in the master build already btw.

Reinhard

**Reply by mavo** [reply]
As always, fast replies. So much appreciated!
Looking forward checking it out. Still on vacation but back tomorrow and will check how it works soon, probably still this weekend.

**Reply by mavo** [reply]
Hello @reinhard-brandstaedter,

not sure if you saw [this](https://github.com/reinhard-brandstaedter/solarflow-control/issues/145#issuecomment-2000412324) in the (closed) issue. Can you maybe check it? I cannot use the new version of sf-control currently.

Regards,
MaVo
